src/main/java/org/jitsi/videobridge/RtpChannel.java

```diff
@@ -1991,6 +1968,121 @@ public RtpChannelTransformEngine getTransformEngine()
         return this.transformEngine;
     }
 
+    /**
+     * Creates the {@code MediaStreamTrack}s from signaling and adds them to the
+     * {@code MediaStream} that is associated to this {@code RtpChannel}.
+     *
+     * @param sources  The <tt>List</tt> of <tt>SourcePacketExtension</tt> that
+     * describes the list of sources of this <tt>RtpChannel</tt> and that is
+     * used as the input in the update of the Sets the <tt>Set</tt> of the SSRCs
+     * that this <tt>RtpChannel</tt> has signaled.
+     * @param sourceGroups
+     */
+    public void setMediaStreamTracks(
+        List<SourcePacketExtension> sources,
+        List<SourceGroupPacketExtension> sourceGroups)
+    {
+        if ((sources == null || sources.isEmpty())
+            && (sourceGroups == null || sourceGroups.isEmpty()))
+        {
+            return;
+        }
+
+        this.setSources(sources); // TODO remove and rely on MSTs.
+        this.setSourceGroups(sourceGroups); // TODO remove and rely on MSTs.
+
+        Map<Long, MediaStreamTrack> tracksBySSRC = new TreeMap<>();
+        Map<Long, long[]> encodings = new TreeMap<>();
+
+        List<Long> freeSSRCs = new ArrayList<>();
+
+        if (sources != null && sources.size() != 0)
+        {
+            for (SourcePacketExtension spe : sources)
+            {
+                freeSSRCs.add(spe.getSSRC());
+            }
+        }
+
+        if (sourceGroups != null && sourceGroups.size() != 0)
+        {
+            for (SourceGroupPacketExtension sgpe : sourceGroups)
+            {
+                if ("sim".equalsIgnoreCase(sgpe.getSemantics())
+                    && sgpe.getSources() != null
+                    && sgpe.getSources().size() >= 2)
+                {
+                    // Every simulcast group determines an mst with a bunch of
+                    // encodings.
+                    MediaStreamTrack track = new MediaStreamTrack();
+
+                    int order = RTPEncoding.BASE_ORDER;
+                    for (SourcePacketExtension spe : sgpe.getSources())
+                    {
+                        long primarySSRC = spe.getSSRC();
+                        freeSSRCs.remove(primarySSRC);
+
+                        long[] encoding = new long[] { -1, -1, -1 };
+                        encoding[0] = primarySSRC;
+                        encoding[1] = order++;
+
+                        encodings.put(primarySSRC, encoding);
+                        tracksBySSRC.put(primarySSRC, track);
+                    }
+                }
+                else if ("fid".equalsIgnoreCase(sgpe.getSemantics())
+                    && sgpe.getSources() != null
+                    && sgpe.getSources().size() == 2)
+                {
+                    Long primarySSRC = sgpe.getSources().get(0).getSSRC();
+                    freeSSRCs.remove(primarySSRC);
+                    Long rtxSSRC = sgpe.getSources().get(1).getSSRC();
```

Comment:
I counted 6 occurrences of `sgpe.getSources()` in this loop. I believe a local variable will make the source code easier to read.